Add hint about synchronizing STDOUT/STDERR in scripted usage

nachtgeist opened the pull request.

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+Hints on scripted usage
+-----------------------
+If you want to run image-bootstrap from a shell script and capture its output
+to STDOUT/STDERR in a log file, be sure to run image-bootstrap in unbuffered
```
Would you mind going lowercase on "STDOUT/STDERR", with code-backticks maybe, e.g. `stdout`/`stderr`? It looks weird uppercase and Wikipedia has is lowercase as well: https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout). Or text version "standard output/error"? I think I like `stdout`/`stderr` best. Two more occurrences below.

nachtgeist replied:
ack

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+to STDOUT/STDERR in a log file, be sure to run image-bootstrap in unbuffered
+mode, e.g.:
+
+`python -u image-bootstrap [params] 2>&1 | tee my.log`
```
Let's either fully inline the command or make it stand out as a block as done elsewhere in this file. I'd recommend this syntax ..

```
------------------------------------------------------------------------------------------
python -u image-bootstrap [params] 2>&1 | tee my.log
------------------------------------------------------------------------------------------
```

.. (with a single blank line before and after) to match the rest of the file.

nachtgeist replied:
Let's go with "elsewhere" style

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+
+The default shebang of image-bootstrap does NOT use `-u`. Without it,
+STDOUT/STDERR won't be synchronized and error output you see in a log file will
+not exactly correspond to preceding/succeding output on STDOUT.
```
I believe "succe[e]ding" is missing an `e`.

nachtgeist replied:
ack

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+
+The default shebang of image-bootstrap does NOT use `-u`. Without it,
+STDOUT/STDERR won't be synchronized and error output you see in a log file will
+not exactly correspond to preceding/succeding output on STDOUT.
```
Can we make that part sound a little less like image-bootstrap would be doing something unusual or broken? It applies to any program with buffered output and buffering is a feature, in some sense, just not in this scenario.

nachtgeist replied:
Sure. No harm intended. Gimme a few days until I have time to work over your comments. Thanks for your input!

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+to STDOUT/STDERR in a log file, be sure to run image-bootstrap in unbuffered
+mode, e.g.:
+
+`python -u image-bootstrap [params] 2>&1 | tee my.log`
```
I wonder what the `stdbuf(1)` version of that command would be. Probably not of more value here but in case you figure it out, please let me know.

nachtgeist replied:
Its man page says also:

>  NOTE: If COMMAND adjusts the buffering of its standard streams ('tee'  does
       for  example)  then  that  will override corresponding changes by 'stdbuf'.

Since python obviously does its own thing, I'm not sure if this could be trusted. Never used stdbuf before and probably never will :)

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
 ------------------------------------------------------------------------------------------
 
 
+[[Scripting]]
```
`README.asciidoc` has a manual TOC — happy to learn about a better way that works! — so `Scripting` will not appear in the TOC… unless you add an entry to it to match.
Or move it to known limitations? (Unsure about that.)

nachtgeist replied:
I'm more of a markdown fan, never used asciidoc before…so I don't know, sorry

nachtgeist commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+[[Scripting]]
+Hints on scripted usage
+-----------------------
+If you want to run image-bootstrap from a shell script and capture its output
```
ack

nachtgeist commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+Hints on scripted usage
+-----------------------
+If you want to run image-bootstrap from a shell script and capture its output
+to STDOUT/STDERR in a log file, be sure to run image-bootstrap in unbuffered
```
What do you mean exactly? I don't follow…

nachtgeist commented on README.asciidoc:
```diff
@@ -369,6 +369,19 @@ command names:
+[[Scripting]]
+Hints on scripted usage
+-----------------------
+If you want to run image-bootstrap from a shell script and capture its output
```
ack

nachtgeist commented:
Just updated the PR with all your comments/requests applied. Also added a note about `set -o pipefail`.
If you have more ideas or further adjustments, just go for it!

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +370,29 @@ command names:
+Hints on using image-bootstrap within a pipe
+--------------------------------------------
+If you want to run *image-bootstrap* in a pipe to capture its output to both
+`stdout/stderr` in a single log file, be sure to run *image-bootstrap* in
```
With my picky hat on, would you mind to apply this change twice:
```diff
--- `stdout/stderr`
+++ `stdout`/`stderr` 
```
I believe that's better semantics.

nachtgeist replied:
ack

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +370,29 @@ command names:
+for passing command parameters only in fairly recent versions. As a consequence,
+`stdout/stderr` won't be synchronized and error output you see in a log file
+will not exactly correspond to preceding/succeeding output on `stdout`. (See
+https://github.com/hartwork/image-bootstrap/issues/71 for more details)
```
Let's not show the full URI but "issue #71", instead. I believe the syntax is `https://github.com/hartwork/image-bootstrap/issues/71[issue #71]`.
After "details", we could add a full stop.

nachtgeist replied:
IMO, nope. Someone might be reading this offline, then being cut off...

hartwork replied:
`https://github.com/hartwork/image-bootstrap/issues/71[issue #71]` is the asciidoc equvalent of markdown `[issue #71](https://github.com/hartwork/image-bootstrap/issues/71)`. so there is more information that the plain link, not less. What you say regarding cut off: was that a misunderstanding? Else please elaborate.

nachtgeist replied:
Being cut off: I was worried about a user having cloned the repo and not being able to resolve the full URL for research.
I guess, the real issue here is between the link text shown and the URL being embedded within the <a href>... 

Like I said: If you'd like more adjustments, go for it. My vacation's almost over and I won't be online as frequent…

hartwork commented on README.asciidoc:
```diff
@@ -369,6 +370,29 @@ command names:
+
+Moreover, if you're using `bash` and you need to keep track of
+*image-bootstrap*'s exit code, be sure to run `set -o pipefail` prior to
+invoking *image-bootstrap*. (Please google `bash pipefail` for more details.)
```
Let's not advertise Google here. Let's use one of the other options:
* `help set | fgrep -A2 pipefail`
* `man bash`
* something like: `section https://www.gnu.org/software/bash/manual/html_node/Pipelines.html[3.2.2 Pipelines] of the https://www.gnu.org/software/bash/manual/html_node/index.html[GNU Bash Reference Manual]`

nachtgeist replied:
ack

hartwork commented:
Merged with 9c7b833b38c83bbbb5c8d7fc18d9e6c80f2e3c43 on top now. Thanks!
